0027-tcp-load-balancer.md

```diff
@@ -0,0 +1,264 @@
+---
+
+# RFD 27 - TCP Load Balancer
+
+## What
+
+A Toy TCP Load Balancer utilising mTLS to authenticate clients and distribute load.
+
+## Why
+
+Coding challenges can give a reasonable expectation of my performance *if* employed by Teleport.
+
+## Details
+
+### Assumptions
+
+- There is only one type of upstream server. This could be changed if using Application Layer Protocol Negotiation extension as part of TLS.
+- Upstream server connection will not include any authntication/authorisation and is a straight tcp connection.
+- Upstream server is expected to implement ping/pong functionality.
+- Upstream servers are grouped and tcp addresses / ports will be hard coded.
+- Authorisation for a client to a server group will be hard coded.
+- The CA, Server keys + cert and client keys + cert will be precomputed using openssl for creation and signing.
+- Client certificate public keys will be used to for authorization data.
+- Maxmium connection count per client will be hard coded to 10.
+
+### Library
+- Single threaded
+    - simplicity
+    - much harder to deadlock
+    - at expense of throughput
+    - could offload to threads or use async
+    - at expense of possible deadlocks, greater complexity.
+
+
+```rust
+struct LoadBalancer
+{
+    clients         : HashMap<String, Client>,
+    server_groups   : HashMap<u32, ServerGroup>,
+}
+```
+
+The main structure will contain a hashmap of clients and server groups.
+Server Groups are an arbitary number of servers that are grouped for security aurhorisation reasons.
+Clients can only connect to one server group, this will be hardcoded as part of the authorisation code.
+The Clients key will be their public key stored as a string.
+
+#### Client Structure
+```rust
+struct Client
+{
+    public_key  : String,
+    connections : Vec<Connection>,
+}
+```
+
+This struct will house all data required to handle all connections (downstream and upstream) on a per client basis.
+It will store the client public key for the client as a means of identification.
+
+```rust
+enum ConnState
+{
+    INIT,
+    UP_CONNECTED,
+    OKAY,
+    UP_DISCONNECT,
+    UP_TIMEOUT,
+    DOWN_DISCONNECT,
+    DOWN_TIMEOUT,
+    DOWN_ENC_ERR,
+}
+```
+
+The state of both connections will be managed by a finite state machine.
+Any state after OKAY will result in the whole connection being dropped, the client will need to retry that connection.
+
+These states could also be used to aid the HealthChecker in marking a node as unhealthy (in the case of UP_DISCONNECTED or UP_TIMEOUT).
+
+The DOWN_ENC_ERR is a catchall for any and all rustls::Stream errors.
+
+```rust
+struct Connection
+{
+    down_stream : TcpStream,
+    up_stream   : TcpStream,
+    tls         : rustls::Stream, 
+    conn_state  : ConnState,
+    up_sg       : u32
+    up_id       : u32,
+}
+```
+
+The connection structure will house all TcpStreams and the rustls::Stream the id of the upstream server and the connection state enum.
+
+#### Client Rate Limiter
```

Comment:
I believe we intend for the rate limiter to be something that limits the amount of new connections over some time window, rather than a hard limit on the amount of existing connections.